router/delivery/http/router_handler.go

```diff
@@ -109,32 +218,64 @@ func (a *RouterHandler) GetOptimalQuote(c echo.Context) (err error) {
 	tokenIn.Denom = chainDenoms[0]
 	tokenOutDenom = chainDenoms[1]
 
-	routerOpts := []domain.RouterOption{}
-
-	// Disable split routes if singleRoute is true
-	if isSingleRoute {
+	var routerOpts []domain.RouterOption
+	if req.SingleRoute {
 		routerOpts = append(routerOpts, domain.WithMaxSplitRoutes(domain.DisableSplitRoutes))
 	}
 
-	quote, err := a.RUsecase.GetOptimalQuote(ctx, tokenIn, tokenOutDenom, routerOpts...)
+	// NOTE:
+	// In general it seems that original idea of just inverting inputs is moving away from the proposed design.
+	// That is because we have to handle different logic for each swap method in different places even with inputs inversion that results in the similar code amount as it would be with specific quote types.
+	// Specifically, we need to implement following changes ( based on original idea ):
+	// - Each pool needs to have added TokenInDenom field
+	// - Each pool needs to have added ChargeTakerFeeExactOut field
+	//
+	// 
+	// Note parameter req.SwapMethod(), this is currently not implemented, but just to demonstrate the idea:
+	// It would make more sense to construct and return specific swap method quote from GetOptimalQuote
+	// Returned domain.Quote's underlying type can be either ExactAmountInQuote or ExactAmountOutQuote
+	//
+	// That would allow to simplify all handling per swap method below.
+	// We could be able to switch between CalculateTokenOutByTokenIn and CalculateTokenInByTokenOut
+	// As a side effect, all quote related methods ( for example custom direct quote ) could benefit for ExactAmountOut logic.
+	quote, err := a.RUsecase.GetOptimalQuote(ctx, *tokenIn, tokenOutDenom, req.SwapMethod(), routerOpts...)
 	if err != nil {
 		return err
 	}
 
 	scalingFactor := oneDec
-	if shouldApplyExponents {
+	if req.ApplyExponents {
 		scalingFactor = a.getSpotPriceScalingFactor(tokenIn.Denom, tokenOutDenom)
 	}
 
-	_, _, err = quote.PrepareResult(ctx, scalingFactor)
+	//  PrepareResult requires swap method to be passed in, because based on that
+	// it needs to calculate takers fee for each pool in the route.
+	_, _, err = quote.PrepareResult(ctx, scalingFactor, req.SwapMethod())
 	if err != nil {
 		return err
 	}
 
 	span.SetAttributes(attribute.Stringer("token_out", quote.GetAmountOut()))
 	span.SetAttributes(attribute.Stringer("price_impact", quote.GetPriceImpact()))
 
-	return c.JSON(http.StatusOK, quote)
+	// If a Quote is swap method specific, we can remove all below
+	if req.IsSwapExactAmountIn() {
+		return c.JSON(http.StatusOK, quote)
+	}
+
+	// For swap exact amount out, we reformat quote: invert amount in and amount out.
+	q := usecase.QuoteExactAmountOut{
+		AmountOut:               quote.GetAmountIn(),
+		AmountIn:                quote.GetAmountOut(),
+		Route:                   quote.GetRoute(),
```

Comment:
Do we need to reformat the route as well?

Reply:
Yes, we need to invert amounts, it's done [here](https://github.com/osmosis-labs/sqs/pull/390/files#diff-8e7615d9ce68d7727b44257f7c7c794396152b9f9d2afdd09505fbdb4880f18aR152)